### PR TITLE
Add tensor.view(dtype) overload support

### PR DESCRIPTION
**Pull request by zasdfgbnm**
In this PR, I added support for `tensor.view(dtype)`, which reinterpret the memory as different types. This overload of `view` is added to the eager mode by myself at https://github.com/pytorch/pytorch/pull/47951. Unfortunately, due to a bug in TorchScript(https://github.com/pytorch/pytorch/issues/47964), it is impossible to enable this overload from the Python side. The only way to use this overload is to manually construct nvFuser IR.

But the main purpose of this PR is not to add a hard-to-use support for a rarely used PyTorch op. The main purpose of this PR is to serve as an incremental step for a bigger story:

In eager mode, `rand_like` for complex tensors is implemented as something like:
```python
def rand_like(t):
    output = torch.empty_like(t)
    output.view_as_real().uniform_()
    return output
```
I want to reproduce the same behavior so that the `randn_like` test passes for complex in `FusionUnaryOps_CUDA`(https://github.com/csarofeen/pytorch/pull/1472/files#diff-682a96c246484fdd4ff073a650e157aa2ffa86c0c0bdd14cf9f4ec7c1b796bcdR4052). To achieve this, I need to add support for `torch.view_as_real`, which views complex tensors of shape `(M1, M2, ...)` into floating pointing tensors of shape `(M1, M2, ..., 2)`. This op creates a view and does not copy memory in eager mode.

I believe adding the `torch.view_as_real` support will be beneficial not only to `rand_like`, but it also provides a mechanism for us to view complex tensors as real number tensors, and reuse existing infrastructure. For example, we could transform a `complex<float>` sum into a `float` sum, if we think this is a good idea.

But adding `torch.view_as_real` sounds like a difficult job to me because it views the tensor as a dtype with different sizes, and it creates a new dim from nowhere. I have no idea how to play with these `TensorDomain`s to make a correct implementation, so I decided to begin with this simpler task: view the tensor as a dtype with the same size. Starting from simpler tasks helps to make my learning curve smoother, and build up the required infrastructure incrementally.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/ops/alias.cpp [hunk 1]
```diff
@@ -52,6 +52,33 @@ TensorView* applyViewTransforms(
...
+  auto inp_domain = x->domain()->getRootDomain();
+  out_domain.reserve(inp_domain.size());
+  for (auto d : inp_domain) {
+    out_domain.push_back(d->clone());
+  }
+  auto out = IrBuilder::create<TensorView>(
+      x->container(),
+      IrBuilder::create<TensorDomain>(
+          out_domain, std::vector<bool>(out_domain.size(), true)),
+      dtype);
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/ops/alias.cpp]: I am not sure if I am doing the correct thing here. The test passes, but I am not confident that the test will find all issues.

File: torch/csrc/jit/codegen/cuda/ops/alias.cpp [hunk 2]
```diff
@@ -52,6 +52,33 @@ TensorView* applyViewTransforms(
 
 } // namespace
 
+TensorView* view(TensorView* x, DataType dtype) {
+  if (x->getDataType() == dtype) {
+    return x;
+  }
+  // TODO: support view(dtype) for dtypes of different size.
+  TORCH_INTERNAL_ASSERT(
+      dataTypeSize(x->getDataType().value()) == dataTypeSize(dtype));
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ops/alias.cpp]: Can you please add an error message?

File: torch/csrc/jit/codegen/cuda/ops/alias.cpp [hunk 3]
```diff
@@ -52,6 +52,33 @@ TensorView* applyViewTransforms(
...
+
+  std::vector<IterDomain*> out_domain;
+  // TODO: should the following line be:
+  //    TensorDomain::noReductions(x->getMaybeRFactorDomain());
+  // ? What does it mean?
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ops/alias.cpp]: Yes, `inp_domain` should be `TensorDomain::noReductions(x->getMaybeRFactorDomain())`.

`TensorDomain::noReductions()` just removes reduction IterDomains. 

`getMaybeRFactorDomain()` returns an rfactor domain if `x` has one. This matters, e.g., `x` is produced by a view operation, where its root domain has the same domain as the input tensor of the view op, and the rfactor domain has the domain after the view. Since this `ViewDtype` op sees `x` as its input, it should look at the post-view domain rather than the pre-view domain.

File: torch/csrc/jit/codegen/cuda/ops/alias.cpp (hunk 1 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ops/alias.cpp]: Looks good except for the above `inp_domain`.

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp [hunk 4]
```diff
@@ -476,6 +476,10 @@ void IrPrinter::handle(const GatherOp* op) {
   os_ << "} )\n";
 }
 
+void IrPrinter::handle(const ViewDtypeOp* top) {
+  indent() << top->out() << " = view.dtype( " << top->in() << " )\n";
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_iostream.cpp]: `top->out()` and `top->in()` don't display their types, so can you please also print the type argument of the dtype view op?

File: torch/csrc/jit/codegen/cuda/runtime/helpers.cu [hunk 5]
```diff
@@ -307,3 +307,25 @@ float pow(float a, int64_t b) {
...
   return pow(a, (double)b);
 }
+
+template <int size, int align = size>
+struct alignas(align) TypelessData {
+  int8_t data[size];
+
+  template <typename T, std::enable_if_t<sizeof(T) == size, int> _ = 0>
+  TypelessData(T x) {
+    memcpy(data, &x, size);
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/runtime/helpers.cu]: @csarofeen recently mentioned he saw some strange results when `memcpy` was used with register values. Just a for-loop copy would be fine too.

**Comment by naoyam** [comment]
@zasdfgbnm I wonder if we would need a new expression, `ViewDtype`. Looks like what it does is basically `reinterpret_cast`, so it may be sufficient to add a new `UnaryOp` type as is the case for `UnaryOpType::Cast`. See `castOp` in arith.cpp as well.

**Comment by naoyam** [comment]
I don't see any red flag except for the setting up of `inp_domain`. As I commented above, I'm not sure if we would need a different expression type. I'm guessing we could do something similar to `castOp`. That said, if this solves an immediate issue, I think we can merge this now and refactor later if necessary.

File: torch/csrc/jit/codegen/cuda/ops/alias.cpp (hunk 3 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ops/alias.cpp]: Thanks for explaining! I have updated this part of code.

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp (hunk 4 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_iostream.cpp]: Added

File: torch/csrc/jit/codegen/cuda/ops/alias.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ops/alias.cpp]: Added

File: torch/csrc/jit/codegen/cuda/runtime/helpers.cu (hunk 5 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/runtime/helpers.cu]: I am not sure about this. If it was a copy of two arrays of the same type, then a for loop could be pretty good, but for the case here, in order to do for loop copy, I will have to use `reinterpret_cast` to treat `&x` as a different type, which is not a recommended way to do type punning according to https://gist.github.com/shafik/848ae25ee209f698763cffee272a58f8#how-do-we-type-pun-correctly
Indeed @ptrblck fixed a bug caused by this strict-aliasing issue https://github.com/pytorch/pytorch/pull/66194 caused by `reinterpret_cast` on ARM. So I have a feeling that `memcpy` should be a safer solution than `reinterpret_cast`.

**Comment by zasdfgbnm** [comment]
Re @naoyam 
> I wonder if we would need a new expression, `ViewDtype`. Looks like what it does is basically `reinterpret_cast`, so it may be sufficient to add a new `UnaryOp` type as is the case for `UnaryOpType::Cast`. See `castOp` in `arith.cpp` as well.

For this PR, yes, the `ViewDtype` is just a unary op like `Cast`. But actually `ViewDtype` should support viewing dtypes of different size. For example:
```python
>>> import torch
>>> a = torch.zeros(128, dtype=torch.int8)
>>> a.view(torch.float)
tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0.])
>>> a.view(torch.float).shape
torch.Size([32])
```
Support for the viewing of dtypes of different sizes is added to the eager mode in PyTorch 1.11, which will release soon. In order to support viewing tensor as another tensor with a dtype of different size in the future, I think it is necessary to have a new expression node type?

File: torch/csrc/jit/codegen/cuda/runtime/helpers.cu (hunk 5 above)
Comment (ptrblck) [reply torch/csrc/jit/codegen/cuda/runtime/helpers.cu]: +1 to @zasdfgbnm's explanation.

> he saw some strange results when memcpy was used with register values

That's also true, and we should investigate it separately. I can check @csarofeen's code again (didn't see any issues so far).

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/runtime/helpers.cu]: Oh, I see. I wonder if it would also mean our vectorized loads and stores with, e.g., `reinterpret_cast<float4*>` were considered unsafe.

https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/runtime/array.cu#L27-L60

Interestingly, he saw an error with `memcpy` and a workaround was to use `reinterpret_cast`.

Comment (ptrblck) [reply torch/csrc/jit/codegen/cuda/runtime/helpers.cu]: > Interestingly, he saw an error with `memcpy` and a workaround was to use `reinterpret_cast`.

Exactly! And so far I don't understand why that's the case. Thanks for sending the code snippet btw.

**Comment by naoyam** [comment]
> Re @naoyam
> 
> > I wonder if we would need a new expression, `ViewDtype`. Looks like what it does is basically `reinterpret_cast`, so it may be sufficient to add a new `UnaryOp` type as is the case for `UnaryOpType::Cast`. See `castOp` in `arith.cpp` as well.
> 
> For this PR, yes, the `ViewDtype` is just a unary op like `Cast`. But actually `ViewDtype` should support viewing dtypes of different size. For example:
> 
> ```python
> >>> import torch
> >>> a = torch.zeros(128, dtype=torch.int8)
> >>> a.view(torch.float)
> tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
>         0., 0., 0., 0., 0., 0., 0., 0.])
> >>> a.view(torch.float).shape
> torch.Size([32])
> ```
> 
> Support for the viewing of dtypes of different sizes is added to the eager mode in PyTorch 1.11, which will release soon. In order to support viewing tensor as another tensor with a dtype of different size in the future, I think it is necessary to have a new expression node type?

I see. Then, yes, it would be reasonable to have its own expression. For supporting different type sizes, it would use the rfactor domain to represent the post-view type. It would be likely to use a similar approach as `view`.

**Comment by zasdfgbnm** [comment]
@naoyam 
> For supporting different type sizes, it would use the rfactor domain to represent the post-view type. It would be likely to use a similar approach as `view`.

Thanks a lot for this information! I was struggling a lot on figuring out how I should achieve this :)

File: torch/csrc/jit/codegen/cuda/runtime/helpers.cu (hunk 5 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/runtime/helpers.cu]: How vectorized loads and stores are done in the PyTorch eager mode? Doesn't it use `reinterpret_cast<vector_type*>`?

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/runtime/helpers.cu]: > How vectorized loads and stores are done in the PyTorch eager mode? Doesn't it use `reinterpret_cast<vector_type*>`?

Unfortunately it does https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/MemoryAccess.cuh#L226 :(

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/runtime/helpers.cu]: The problem only happens when a tensor is loaded and stored in an inconsistent way, i.e., a vectorized store followed by a non-vectorized load, or a non-vectorized store followed by a vectorized load. And it only matters when it happens in a single CUDA kernel. I wonder when such patterns would appear. Unlikely?

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/runtime/helpers.cu]: When memcpy was used for our vectorized loads I think nvrtc had a hard time understanding the size in registers that was required for some reason. I think it was effectively not able to understand the side effects from the registers being referenced from the ptr + size. Reinterpret cast seemed to effectively indicate to the compiler "this is the set of registers I'm expecting". However, this is more of a hunch than a validated cause.

**Comment by csarofeen** [comment]
This is treading in new territory, but I think Naoya is right. The concept of RFactor is that we have a transformation in the overall root dimension of tensors that we need to track. View and rfactor for reductions has a set of transformations (split/merge) from a root domain to the rfactor domain. The Root domain maps to the producers of a given tensor, the rfactor domain maps to the consumers of a given tensor.

The mappings allow us to do things like replay mappings across expressions so we can transform tensors consistently, and analyze their relationship. The reason this is treading in slightly new territory is that you'd have an extra dimension that has no mapping from root domain to rfactor. I think the more challenging aspect of this (than building out the op) is how do we support replays and mappings across the operation. That though can be left as follow ups.

**Comment by csarofeen** [comment]
Some more notes, there will also be some challenge here where the extraneous dimension shouldn't be indexed because it's really like a vectorized load. i.e. if we have inp [4, 4] and output of [4, 4, 2], that 2 shouldn't be indexed as you're effectively doing a vectorized store in the output. This shouldn't be too difficult, just don't know if we should always mark that dimension as vectorized in lowering.

File: torch/csrc/jit/codegen/cuda/lower_fusion_simplifier.cpp [hunk 6]
```diff
@@ -91,6 +91,15 @@ class UnaryOpInserter : private kir::ExprMutator {
         gop, IrBuilder::create<UnaryOp>(container, UnaryOpType::Set, out, in));
   }
 
+  void handle(ViewDtypeOp* vop) final {
+    auto out = vop->out();
+    auto in = vop->in();
+    auto container = out->container();
+    registerReplace(
+        vop,
+        IrBuilder::create<UnaryOp>(container, UnaryOpType::EraseType, out, in));
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_fusion_simplifier.cpp:100]: I'm uncertain if we want to replace this op before lowering. We might consider reverting this in some of the follow up work.

File: torch/csrc/jit/codegen/cuda/root_domain_map.h [hunk 7]
```diff
@@ -397,6 +397,10 @@ class TORCH_CUDA_CU_API ComputeAtRootDomainMapBuilder
     mapPointwiseOrReductionOp(op);
   }
 
+  void handle(ViewDtypeOp* op) override {
+    mapPointwiseOrReductionOp(op);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/root_domain_map.h:401]: This will be important to fill out manually when modifying the rfactor domain of the producer/ root domain of the output..

File: torch/csrc/jit/codegen/cuda/runtime/helpers.cu (hunk 5 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/runtime/helpers.cu]: Basically my opinion here is whatever works. If you want to see where it broke us you can revert the commenting of the memcpy in the runtime files for vectorized load/store and run the c++ tests.

**Review by csarofeen — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/runtime/helpers.cu (hunk 5 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/runtime/helpers.cu]: Marking this as resolved. Per offline discussion, we should prefer `reinterpret_cast`. And I also updated my PR using `reinterpret_cast`.
